contracts/Registry/TokenCuratedRegistry.sol

```diff
@@ -1,59 +1,107 @@
+  constructor(ERC20 _token, uint _minStake, uint _applicationPeriod, IChallengeFactory _challengeFactory)
+  StakedRegistry(_token, _minStake)
+  public {
+    applicationPeriod = _applicationPeriod;
+    challengeFactory = _challengeFactory;
+  }
 
-  modifier itemNotLockedInChallenge(bytes32 id) {
-    require(!inChallengePhase(id));
-    _;
+  // Adds an item to the `items` mapping, transfers token stake from msg.sender, and locks
+  // the item from removal until now + applicationPeriod.
+  function add(bytes32 data) public returns (bytes32 id) {
+    id = super.add(data);
+    unlockTimes[id] = now.add(applicationPeriod);
   }
 
-  mapping(bytes32 => ItemCurationData) itemsCurationData;
+  // Removes an item from the `items` mapping, and deletes challenge state. Requires that
+  // there is not an active or passed challenge for this item. OwnedItemRegistry.remove
+  // requires that this is called by the item owner. LockableItemRegistry.remove requires
+  // that the item is not locked.
+  function remove(bytes32 id) public {
+    if (challengeExists(id)) {
+      require(challengeEnded(id) && !challengePassed(id));
+    }
+    delete challenges[id];
+    super.remove(id);
+  }
 
-  struct ItemCurationData {
-    uint applicationExpiry;
-    bool whitelisted;
-    address challengeAddress;
-    bool challengeResolved;
+  // Creates a new challenge for an item.
+  // Requires that the item exists, and that there is no existing challenge for the item.
+  // Requires msg.sender (the challenger) to match the owner's stake by transferring to
+  // this contract. The challenger's and owner's stake is transferred to the newly created
+  // challenge contract.
+  function challenge(bytes32 id) public {
+    require(!challengeExists(id));
+    require(token.transferFrom(msg.sender, this, minStake));
+    address challenge = challengeFactory.create(this, msg.sender, owners[id]);
+    require(token.transfer(challenge, minStake.mul(2)));
   }
 
-  constructor (
-    ERC20 _token,
-    uint _minStake,
-    uint _applicationPeriod,
-    ChallengeFactory _challengeFactory
-  ) public
-    StakedRegistry(_token, _minStake)
-  {
-    applicationPeriod = _applicationPeriod;
-    challengeFactory = _challengeFactory;
+  // Handles transfer of reward after a challenge has ended. Requires that there
+  // is an ended challenge for the item.
+  function resolveChallenge(bytes32 id) public {
+    require(challengeEnded(id));
+    uint reward = challenges[id].reward();
+    require(token.transferFrom(challenges[id], this, reward));
+    if (challengePassed(id)) {
+      // if the challenge passed, reward the challenger (via token.transfer) and remove
+      // the item.
+      require(token.transfer(challenges[id].challenger(), reward));
+      delete ownerStakes[id];
+      super.remove(id);
+    } else {
+      // if the challenge failed, reward the applicant (by adding to their staked balance)
+      ownerStakes[id] = ownerStakes[id].add(reward).sub(minStake);
```

Comment:
if minStake has changed since the challenge began, our accounting is going to be off here. We tracked this in the original version in the challengeStruct. I'm wondering how much some extra data storage would help with little things like this.

Reply:
@emaG3m that's true, but there's no way to change minStake in this implementation. I think we should deal with this issue when we add a way to change minStake